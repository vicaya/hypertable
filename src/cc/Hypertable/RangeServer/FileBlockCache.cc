/** -*- c++ -*-
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include <cassert>
#include <iostream>

#include "FileBlockCache.h"

using namespace Hypertable;
using std::pair;

atomic_t FileBlockCache::ms_next_file_id = ATOMIC_INIT(0);

FileBlockCache::~FileBlockCache() {
  for (BlockCache::const_iterator iter = m_cache.begin();
       iter != m_cache.end(); ++iter)
    delete [] (*iter).block;
}

bool
FileBlockCache::checkout(int file_id, uint32_t file_offset, uint8_t **blockp,
                         uint32_t *lengthp) {
  ScopedLock lock(m_mutex);
  HashIndex &hash_index = m_cache.get<1>();
  HashIndex::iterator iter;

  m_accesses++;
  int64_t key = ((int64_t)file_id << 32) | file_offset;

  if ((iter = hash_index.find(key)) == hash_index.end())
    return false;

  BlockCacheEntry entry = *iter;
  entry.ref_count++;

  hash_index.erase(iter);

  pair<Sequence::iterator, bool> insert_result = m_cache.push_back(entry);
  assert(insert_result.second);

  *blockp = (*insert_result.first).block;
  *lengthp = (*insert_result.first).length;

  m_hits++;
  return true;
}


void FileBlockCache::checkin(int file_id, uint32_t file_offset) {
  ScopedLock lock(m_mutex);
  HashIndex &hash_index = m_cache.get<1>();
  HashIndex::iterator iter;
  int64_t key = ((int64_t)file_id << 32) | file_offset;

  iter = hash_index.find(key);

  assert(iter != hash_index.end() && (*iter).ref_count > 0);

  hash_index.modify(iter, DecrementRefCount());
}


bool
FileBlockCache::insert_and_checkout(int file_id, uint32_t file_offset,
                                    uint8_t *block, uint32_t length) {
  ScopedLock lock(m_mutex);
  HashIndex &hash_index = m_cache.get<1>();
  int64_t key = ((int64_t)file_id << 32) | file_offset;

  if (length > m_limit || hash_index.find(key) != hash_index.end())
    return false;

  if (m_available < length)
    make_room(length);

  if (m_available < length)
    return false;

  BlockCacheEntry entry(file_id, file_offset);
  entry.block = block;
  entry.length = length;
  entry.ref_count = 1;

  pair<Sequence::iterator, bool> insert_result = m_cache.push_back(entry);
  assert(insert_result.second);

  m_available -= length;

  return true;
}


bool FileBlockCache::contains(int file_id, uint32_t file_offset) {
  ScopedLock lock(m_mutex);
  HashIndex &hash_index = m_cache.get<1>();
  int64_t key = ((int64_t)file_id << 32) | file_offset;
  m_accesses++;

  if (hash_index.find(key) != hash_index.end()) {
    m_hits++;
    return true;
  }
  else
    return false;
}


void FileBlockCache::increase_limit(int64_t amount) {
  ScopedLock lock(m_mutex);
  int64_t adjusted_amount = amount;
  if ((m_max_memory-m_limit) < amount)
    adjusted_amount = m_max_memory - m_limit;
  m_limit += adjusted_amount;
  m_available += adjusted_amount;
}


int64_t FileBlockCache::decrease_limit(int64_t amount) {
  ScopedLock lock(m_mutex);
  int64_t memory_freed = 0;
  if (m_available < amount) {
    if (amount > (m_limit - m_min_memory))
      amount = m_limit - m_min_memory;
    memory_freed = make_room(amount);
    if (m_available < amount)
      amount = m_available;
  }
  m_available -= amount;
  m_limit -= amount;
  return memory_freed;
}


int64_t FileBlockCache::make_room(int64_t amount) {
  BlockCache::iterator iter = m_cache.begin();
  int64_t amount_freed = 0;
  while (iter != m_cache.end()) {
    if ((*iter).ref_count == 0) {
      m_available += (*iter).length;
      amount_freed += (*iter).length;
      delete [] (*iter).block;
      iter = m_cache.erase(iter);
      if (m_available >= amount)
	break;
    }
    else
      ++iter;
  }
  return amount_freed;
}

void FileBlockCache::get_stats(uint64_t *max_memoryp, uint64_t *available_memoryp,
                               uint64_t *accessesp, uint64_t *hitsp) {
  ScopedLock lock(m_mutex);
  *max_memoryp = m_limit;
  *available_memoryp = m_available;
  *accessesp = m_accesses;
  *hitsp = m_hits;
}
