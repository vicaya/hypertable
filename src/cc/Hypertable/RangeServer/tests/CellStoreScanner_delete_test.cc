/** -*- c++ -*-
 * Copyright (C) 2009 Sanjit Jhala (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include "Common/Init.h"
#include "Common/DynamicBuffer.h"
#include "Common/FileUtils.h"
#include "Common/InetAddr.h"
#include "Common/System.h"
#include "Common/Usage.h"

#include <iostream>
#include <fstream>

#include "AsyncComm/ConnectionManager.h"

#include "DfsBroker/Lib/Client.h"

#include "Hypertable/Lib/Key.h"
#include "Hypertable/Lib/Schema.h"
#include "Hypertable/Lib/SerializedKey.h"

#include "../CellStoreV5.h"
#include "../FileBlockCache.h"
#include "../Global.h"

#include <cstdlib>

using namespace Hypertable;
using namespace std;

namespace {
  const char *usage[] = {
    "usage: CellStoreScanner_test",
    "",
    "  This program tests for the proper functioning of the CellStore",
    "  scanner.  It creates a dummy cell store and then repeatedly scans",
    "  it with different ranges",
    (const char *)0
  };
  const char *schema_str =
  "<Schema>\n"
  "  <AccessGroup name=\"default\">\n"
  "    <ColumnFamily id=\"1\">\n"
  "      <Name>tag</Name>\n"
  "    </ColumnFamily>\n"
  "  </AccessGroup>\n"
  "</Schema>";

  const char *words[] = {
    "prolif",
    "yonsid",
    "testam",
    "hypost",
    "scagli",
    "bandym",
    "protei",
    "paleot",
    "hetero",
    "undeli",
    "Megach",
    "hepari",
    "salama",
    "stereo",
    "favora",
    "aeroli",
    "ovopyr",
    "persec",
    "fondle",
    "Diplod",
    "federa",
    "folkfr",
    "doorli",
    "healin",
    "rugosi",
    "underh",
    "consum",
    "overle",
    "cumins",
    "Hannib",
    "etiolo",
    "unprop",
    "Julian",
    "unsnar",
    "tarril",
    "bratti",
    "tettig",
    "gariba",
    "spermo",
    "palmic",
    "Lauren",
    "basipo",
    "peculi",
    "hetero",
    "unders",
    "untoot",
    "propro",
    "ostrac",
    "Felapt",
    "alloty",
    "Macrur",
    "verdel",
    "semici",
    "humbug",
    "Easter",
    "epacri",
    "holopt",
    "unguid",
    "Nummul",
    "fondak",
    "protog",
    "microm",
    "blight",
    "Guauae",
    "bradya",
    "typhom",
    "ducato",
    "trapez",
    "vipres",
    "coilin",
    "depanc",
    "gnatho",
    "micasi",
    "garnet",
    "thrack",
    "humora",
    "Philis",
    "tropic",
    "cloyso",
    "oxypro",
    "uphols",
    "therol",
    "seawor",
    "nigrif",
    "Calyst",
    "bulimi",
    "leptot",
    "protiu",
    "schrei",
    "laemod",
    "Melast",
    "nonalu",
    "nonsen",
    "chatoy",
    "Pomace",
    "implor",
    "syngen",
    "infide",
    "opiany",
    "archit",
    "unwife",
    "nonthi",
    "cataco",
    "dispul",
    "formal",
    "afters",
    "indica",
    "turnca",
    "feldsp",
    "uralit",
    "suidia",
    "hydrol",
    "parapo",
    "cavern",
    "ammoni",
    "propit",
    "overan",
    "fringe",
    "planet",
    "gangli",
    "rehand",
    "lucidn",
    "anthra",
    "Palaeo",
    "murgav",
    "Billji",
    "thingl",
    "hundre",
    "spheru",
    "worser",
    "phalli",
    "blaene",
    "Achern",
    "Acroce",
    "forese",
    "phytoo",
    "palaeo",
    "enfudd",
    "Laemod",
    "urolit",
    "presyn",
    "harmon",
    "preade",
    "bucket",
    "redoub",
    "submax",
    "sociog",
    "epidid",
    "capsul",
    "interc",
    "unders",
    "chryso",
    "couser",
    "untran",
    "argill",
    "Alkora",
    "antisy",
    "undisp",
    "indire",
    "clashy",
    "uncomm",
    "glaieu",
    "electr",
    "undere",
    "preten",
    "tuberc",
    "terato",
    "docume",
    "Manche",
    "facien",
    "lethol",
    "usitat",
    "othelc",
    "primog",
    "gypsog",
    "earthb",
    "backst",
    "cloudi",
    "uncouc",
    "Panaya",
    "utricu",
    "person",
    "mediov",
    "transc",
    "acetab",
    "polycy",
    "skunkd",
    "prepre",
    "delftw",
    "Acanth",
    "frostp",
    "blackg",
    "tetraz",
    "Olympi",
    "lubrif",
    "grazab",
    "Schist",
    "divers",
    "unconc",
    "nonven",
    "overbl",
    "vivise",
    "visual",
    "postin",
    "catach",
    "tribal",
    "digyni",
    "outhor",
    "Tachin",
    "Syriol",
    "histor",
    "Amoreu",
    "coachm",
    "absent",
    "stibic",
    "subseq",
    "nonoec",
    "caraco",
    "Caripu",
    "Uranic",
    "fletch",
    "acedia",
    "temera",
    "beadle",
    "bancal",
    "mordic",
    "superd",
    "Polyne",
    "Interl",
    "diapho",
    "contin",
    "wastem",
    "cubele",
    "Chorda",
    "unsati",
    "prefin",
    "Amanda",
    "Micros",
    "nonrep",
    "corpus",
    "precas",
    "uncrib",
    "delayf",
    "carlin",
    "dilogy",
    "gravim",
    "unstop",
    "shorem",
    "speakl",
    "inenar",
    "antiho",
    "benzol",
    "inoppo",
    "decrus",
    "compos",
    "Stenop",
    "Rhapis",
    "youngl",
    "laeotr",
    "cannon",
    "nonute",
    "phyllo",
    "ascidi",
    "berate",
    "holoqu",
    "analep",
    "kynuri",
    "conver",
    "overfa",
    "pigflo",
    "suprap",
    "Mattap",
    "citabl",
    "urocer",
    "altern",
    "Subosc",
    "dietet",
    "spiffi",
    "perica",
    "placen",
    "circum",
    "aeroph",
    "harmon",
    "hodder",
    "morphe",
    "marmot",
    "bechir",
    "superc",
    "undome",
    "noncen",
    "Teuton",
    "Uragog",
    "scribi",
    "endodo",
    "praeta",
    "smirkl",
    "Redemp",
    "superr",
    "haplop",
    "poster",
    "chills",
    "Cacaja",
    "unpain",
    "concep",
    "unprac",
    "expunc",
    "ticket",
    "Boulan",
    "lamini",
    "treaty",
    "smokis",
    "straig",
    "hypocr",
    "overle",
    "defini",
    "Dalmat",
    "straig",
    "scoldi",
    "ulster",
    "prevol",
    "redesp",
    "polyhy",
    "unhang",
    "habita",
    "unscra",
    "millif",
    "befume",
    "Panhel",
    "malaci",
    "omnipa",
    "relent",
    "rockal",
    "Royena",
    "Varang",
    "cytoge",
    "superc",
    "pluris",
    "skedad",
    "recons",
    "interp",
    "unclas",
    "infruc",
    "folded",
    "bronch",
    "unlawf",
    "bridge",
    "thinka",
    "Sudani",
    "singab",
    "triflo",
    "slumwi",
    "Aepyce",
    "muskro",
    "eucras",
    "Heracl",
    "ungirt",
    "tinker",
    "supple",
    "martel",
    "tympan",
    "octona",
    "neebor",
    "semime",
    "theodo",
    "remedi",
    "unsucc",
    "agangl",
    "labial",
    "Termit",
    "irrevo",
    "Schope",
    "expans",
    "propro",
    "theelo",
    "unslep",
    "greyne",
    "palust",
    "eventl",
    "danali",
    "bisymm",
    "Opisth",
    "outben",
    "agrono",
    "schizo",
    "retake",
    "subdeb",
    "plotte",
    "palsgr",
    "Gonyst",
    "stickf",
    "pretra",
    "muffed",
    "statut",
    "hinoid",
    "logist",
    "centra",
    "stepch",
    "forema",
    "oometr",
    "nubige",
    "undimi",
    "deutop",
    (const char *)0
  };

  size_t display_scan(CellListScannerPtr &scanner, ostream &out) {
    Key key_comps;
    ByteString bsvalue;
    size_t count = 0;
    while (scanner->get(key_comps, bsvalue)) {
      out << key_comps << "\n";
      count++;
      scanner->forward();
    }
    out << flush;
    return count;
  }
}


int main(int argc, char **argv) {
  try {
    struct sockaddr_in addr;
    ConnectionManagerPtr conn_mgr;
    DfsBroker::ClientPtr client;
    CellStorePtr cs;
    std::ofstream out("CellStoreScanner_delete_test.output");
    String delete_row = "delete_row";
    String delete_cf  = "delete_cf";
    String delete_row_cf = "delete_row_cf";
    String delete_none = "delete_none";
    String delete_large = "delete_large";
    String insert = "insert";

    Config::init(argc, argv);

    if (Config::has("help"))
      Usage::dump_and_exit(usage);

    ReactorFactory::initialize(2);

    uint16_t port = Config::properties->get_i16("DfsBroker.Port");

    InetAddr::initialize(&addr, "localhost", port);

    conn_mgr = new ConnectionManager();
    Global::dfs = new DfsBroker::Client(conn_mgr, addr, 15000);

    // force broker client to be destroyed before connection manager
    client = (DfsBroker::Client *)Global::dfs.get();

    if (!client->wait_for_connection(15000)) {
      HT_ERROR("Unable to connect to DFS");
      return 1;
    }

    Global::block_cache = new FileBlockCache(100000LL, 100000LL);
    Global::memory_tracker = new MemoryTracker(Global::block_cache);

    String testdir = "/CellStoreScanner_delete_test";
    client->mkdirs(testdir);

    SchemaPtr schema = Schema::new_instance(schema_str, strlen(schema_str));
    if (!schema->is_valid()) {
      HT_ERRORF("Schema Parse Error: %s", schema->get_error_string());
      exit(1);
    }

    String csname = testdir + "/cs0";
    PropertiesPtr cs_props = new Properties();
    // make sure blocks are small so only one key value pair fits in a block
    cs_props->set("blocksize", uint32_t(32));
    cs = new CellStoreV5(Global::dfs.get(), schema.get());
    HT_TRY("creating cellstore", cs->create(csname.c_str(), 24000, cs_props));

    DynamicBuffer dbuf(512000);
    String row;
    String qualifier;
    SerializedKey serkey;
    int64_t timestamp = 1;
    std::vector<SerializedKey> serkeyv;
    std::vector<Key> keyv;
    Key key;
    ScanContextPtr scan_ctx;
    String value="0";
    uint8_t valuebuf[128];
    uint8_t *uptr;
    ByteString bsvalue;
    int64_t num_deletes=0;

    uptr = valuebuf;
    Serialization::encode_vi32(&uptr,value.length());
    strcpy((char *)uptr, value.c_str());
    bsvalue.ptr = valuebuf;


    // test delete logic
    {
      // delete row
      serkey.ptr = dbuf.ptr;
      row = delete_row;
      qualifier = insert;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1,
                            qualifier.c_str(), timestamp, timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_ROW, row.c_str(), 0, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      // delete column family
      serkey.ptr = dbuf.ptr;
      row = delete_cf;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, qualifier.c_str(), timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_COLUMN_FAMILY, row.c_str(), 1, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      // delete row & column family
      serkey.ptr = dbuf.ptr;
      row = delete_row_cf;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, qualifier.c_str(), timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_ROW, row.c_str(), 0, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, qualifier.c_str(), timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_COLUMN_FAMILY, row.c_str(), 1, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, qualifier.c_str(), timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_ROW, row.c_str(), 0, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_COLUMN_FAMILY, row.c_str(), 1, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      // delete none
      serkey.ptr = dbuf.ptr;
      row = delete_none;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, qualifier.c_str(), timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);


      // delete large
      serkey.ptr = dbuf.ptr;
      row = delete_large;
      create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, qualifier.c_str(), timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);

      size_t wordi = 0;
      String word;
      while (dbuf.fill() < 140000) {
        serkey.ptr = dbuf.ptr;
        if (words[wordi] == 0)
          wordi = 0;
        word = words[wordi++];
        create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, word.c_str(), timestamp,
                              timestamp);
        timestamp++;
        serkeyv.push_back(serkey);
      }

      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_COLUMN_FAMILY, row.c_str(), 1, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;

      while (dbuf.fill() < 280000) {
        serkey.ptr = dbuf.ptr;
        if (words[wordi] == 0)
          wordi = 0;
        word = words[wordi++];
        create_key_and_append(dbuf, FLAG_INSERT, row.c_str(), 1, word.c_str(), timestamp,
                              timestamp);
        timestamp++;
        serkeyv.push_back(serkey);
      }
      serkey.ptr = dbuf.ptr;
      create_key_and_append(dbuf, FLAG_DELETE_ROW, row.c_str(), 0, "", timestamp,
                            timestamp);
      timestamp++;
      serkeyv.push_back(serkey);
      num_deletes++;
    }

    sort(serkeyv.begin(), serkeyv.end());

    keyv.reserve( serkeyv.size() );

    out << "[baseline]\n";
    for (size_t i=0; i<serkeyv.size(); i++) {
      key.load( serkeyv[i] );
      cs->add(key, bsvalue);
      if (delete_large.compare(key.row) || key.flag == FLAG_DELETE_ROW ||
          key.flag == FLAG_DELETE_COLUMN_FAMILY || !insert.compare(key.column_qualifier))
        out << key << "\n";
      keyv.push_back(key);
    }

    TableIdentifier table_id("0");
    cs->finalize(&table_id);

    RangeSpec range;
    range.start_row = "";
    range.end_row = Key::END_ROW_MARKER;

    ScanSpecBuilder ssbuilder;
    String column;

    CellListScannerPtr scanner;

    /**
     * Test deletes
     */

    out << "[delete-row]\n";
    ssbuilder.clear();
    row = delete_row;
    column = (String)"tag:" + qualifier;
    ssbuilder.add_cell(row.c_str(), column.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    ssbuilder.clear();
    ssbuilder.add_cell_interval(row.c_str(),"tag:a", true,
        row.c_str(), "tag:z", true);
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    ssbuilder.clear();
    ssbuilder.add_row(row.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    out << "[delete-cf]\n";
    ssbuilder.clear();
    row = delete_cf;
    column = (String)"tag:" + qualifier;
    ssbuilder.add_cell(row.c_str(), column.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    ssbuilder.clear();
    ssbuilder.add_cell_interval(row.c_str(),"tag:a", true,
        row.c_str(), "tag:z", true);
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    ssbuilder.clear();
    ssbuilder.add_cell_interval(row.c_str(),"tag", true,
        row.c_str(), "tag:z", true);
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    ssbuilder.clear();
    ssbuilder.add_row(row.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);


    out << "[delete-row-cf]\n";
    ssbuilder.clear();
    row = delete_row_cf;
    column = (String)"tag:" + qualifier;
    ssbuilder.add_cell(row.c_str(), column.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    ssbuilder.clear();
    ssbuilder.add_cell_interval(row.c_str(),"tag", true,
        row.c_str(), "tag", true);
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    out << "[delete-none]\n";
    ssbuilder.clear();
    row = delete_none;
    column = (String) "tag:"+qualifier;
    ssbuilder.add_cell(row.c_str(), column.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                                   schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    out << "[delete-large]\n";
    ssbuilder.clear();
    row = delete_large;
    column = (String)"tag:" + qualifier;
    ssbuilder.add_cell(row.c_str(), column.c_str());
    scan_ctx = new ScanContext(TIMESTAMP_MAX, &(ssbuilder.get()), &range,
                               schema);
    scanner = cs->create_scanner(scan_ctx);
    display_scan(scanner, out);

    int64_t delete_count = boost::any_cast<int64_t>(cs->get_trailer()->get("delete_count"));
    out << "trailer.delete_count= " << delete_count << "\n";
    if (delete_count != num_deletes) {
      out << "Expected " << num_deletes << " deletes in CellStore, but trailer.delete_count="
          << delete_count << endl;
      return 1;
    }

    out << flush;
    String cmd_str = "diff CellStoreScanner_delete_test.output "
                     "CellStoreScanner_delete_test.golden";
    if (system(cmd_str.c_str()) != 0)
      return 1;

    // close cell store
    scanner = 0;
    cs = 0;

    client->rmdir(testdir);
  }
  catch (Exception &e) {
    HT_ERROR_OUT << e << HT_END;
    return 1;
  }
  catch (...) {
    HT_ERROR_OUT << "unexpected exception caught" << HT_END;
    return 1;
  }
  return 0;
}

