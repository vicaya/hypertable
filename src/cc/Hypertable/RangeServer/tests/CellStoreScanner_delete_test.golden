[baseline]
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='' ts=4 rev=4 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='insert' ts=3 rev=3
control=(REV|TS|SHARED) row='delete_large' ts=8753 rev=8753 DELETE_ROW
control=(REV|TS|SHARED) row='delete_large' family=1 qualifier='' ts=4378 rev=4378 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_large' family=1 qualifier='insert' ts=13 rev=13
control=(REV|TS|SHARED) row='delete_none' family=1 qualifier='insert' ts=12 rev=12
control=(REV|TS|SHARED) row='delete_row' ts=2 rev=2 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row' family=1 qualifier='insert' ts=1 rev=1
control=(REV|TS|SHARED) row='delete_row_cf' ts=10 rev=10 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row_cf' ts=6 rev=6 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='' ts=11 rev=11 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='' ts=8 rev=8 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=9 rev=9
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=7 rev=7
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=5 rev=5
[delete-row]
control=(REV|TS|SHARED) row='delete_row' ts=2 rev=2 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row' family=1 qualifier='insert' ts=1 rev=1
control=(REV|TS|SHARED) row='delete_row' ts=2 rev=2 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row' family=1 qualifier='insert' ts=1 rev=1
control=(REV|TS|SHARED) row='delete_row' ts=2 rev=2 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row' family=1 qualifier='insert' ts=1 rev=1
[delete-cf]
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='' ts=4 rev=4 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='insert' ts=3 rev=3
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='' ts=4 rev=4 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='insert' ts=3 rev=3
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='' ts=4 rev=4 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='insert' ts=3 rev=3
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='' ts=4 rev=4 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_cf' family=1 qualifier='insert' ts=3 rev=3
[delete-row-cf]
control=(REV|TS|SHARED) row='delete_row_cf' ts=10 rev=10 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row_cf' ts=6 rev=6 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='' ts=11 rev=11 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='' ts=8 rev=8 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=9 rev=9
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=7 rev=7
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=5 rev=5
control=(REV|TS|SHARED) row='delete_row_cf' ts=10 rev=10 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row_cf' ts=6 rev=6 DELETE_ROW
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='' ts=11 rev=11 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='' ts=8 rev=8 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=9 rev=9
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=7 rev=7
control=(REV|TS|SHARED) row='delete_row_cf' family=1 qualifier='insert' ts=5 rev=5
[delete-none]
control=(REV|TS|SHARED) row='delete_none' family=1 qualifier='insert' ts=12 rev=12
[delete-large]
control=(REV|TS|SHARED) row='delete_large' ts=8753 rev=8753 DELETE_ROW
control=(REV|TS|SHARED) row='delete_large' family=1 qualifier='' ts=4378 rev=4378 DELETE_COLUMN_FAMILY
control=(REV|TS|SHARED) row='delete_large' family=1 qualifier='insert' ts=13 rev=13
trailer.delete_count= 8
